core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -540,6 +540,10 @@ private RexNode simplifyCoalesce(RexCall call) {
     case 0:
       return rexBuilder.makeNullLiteral(call.type);
     case 1:
+      // ensure type consistency, especially nullability.
+      if (!call.getType().equals(operands.get(0).getType())) {
```

Comment:
> Cause we use RexSimplify.simplify directly in rules but simplifyPreservingType is only a tool method for now

Why do you think `simplifyCoalesce` should preserve types?

Why bother with types of **internal** expressions if the only thing Volcano cares is the type of OVERALL expression?

Making internal types (that is types of nested `coalesce` expressions) simpler enables faster operation, and it might open more possibilities for optimizations.

Reply:
Okey, got you idea, then how can we use the Rules that use `simplifyCoalesce` during volcano planning. i.e. ReduceExpressionRule. If we have a better way to support nullability for rules them selfs, it may be a better way.

Reply:
AFAIK `simplifyCoalesce` is a `private` method, and it cannot be used in `ReduceExpressionRule`.

Could you please clarify?

`ReduceExpressionRule` does have `matchNullability`, so either it does not work or it is misused.

Reply:
It is `simplify`, `ReduceExpressionRule` does have a flag to control nullability for` ExprSimplifier` but not `RexSimplify`, that is the case.